File: test/language/explicit_parameters/java/explicit_parameters/explicit_blob_param/ExplicitBlobParamTest.java
package explicit_parameters.explicit_blob_param;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import explicit_parameters.ExplicitParametersDb;

import zserio.runtime.ZserioError;

public class ExplicitBlobParamTest
{
    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @BeforeEach
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new ExplicitParametersDb(file.toString());
        database.createSchema();
    }

    @AfterEach
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable blobParamTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        blobParamTable.write(writtenRows);

        final BlobParamTableParameterProvider parameterProvider = new BlobParamTableParameterProvider();
        final List<BlobParamTableRow> readRows = blobParamTable.read(parameterProvider);
        checkBlobParamTableRows(writtenRows, readRows);
    }

    @Test
    public void readWithCondition() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable blobParamTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        blobParamTable.write(writtenRows);

        final BlobParamTableParameterProvider parameterProvider = new BlobParamTableParameterProvider();
        final String condition = "name='Name1'";
        final List<BlobParamTableRow> readRows = blobParamTable.read(parameterProvider, condition);
        assertEquals(1, readRows.size());

        final int expectedRowNum = 1;
        final BlobParamTableRow readRow = readRows.get(0);
        checkBlobParamTableRow(writtenRows.get(expectedRowNum), readRow);
    }

    @Test
    public void update() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable blobParamTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        blobParamTable.write(writtenRows);

        final int updateRowId = 3;
        final BlobParamTableRow updateRow = createBlobParamTableRow(updateRowId, "UpdatedName");
        final String updateCondition = "id=" + updateRowId;
        blobParamTable.update(updateRow, updateCondition);

        final BlobParamTableParameterProvider parameterProvider = new BlobParamTableParameterProvider();
        final List<BlobParamTableRow> readRows = blobParamTable.read(parameterProvider, updateCondition);
        assertEquals(1, readRows.size());

        final BlobParamTableRow readRow = readRows.get(0);
        checkBlobParamTableRow(updateRow, readRow);
    }

    private static class BlobParamTableParameterProvider implements BlobParamTable.ParameterProvider
    {
        @Override
        public Header getHeaderParam(ResultSet resultSet)
        {
            return headerParam;
        }

        @Override
        public Header getBlob(ResultSet resultSet)
        {
            return blob;
        }

        Header headerParam = new Header(BLOB_PARAM_TABLE_HEADER_COUNT);
        Header blob = new Header(BLOB_PARAM_TABLE_BLOB_COUNT);
    }

    private static void fillBlobParamTableRows(List<BlobParamTableRow> rows)
    {
        for (int id = 0; id < NUM_BLOB_PARAM_TABLE_ROWS; ++id)
        {
            rows.add(createBlobParamTableRow(id, "Name" + id));
        }
    }

    private static BlobParamTableRow createBlobParamTableRow(long id, String name)
    {
        final BlobParamTableRow row = new BlobParamTableRow();

        row.setId(id);
        row.setName(name);

        final Header header = new Header(BLOB_PARAM_TABLE_HEADER_COUNT);

        final byte[] values1 = new byte[BLOB_PARAM_TABLE_HEADER_COUNT];
        for (int i = 0; i < values1.length; ++i)
            values1[i] = (byte)id;
        final TestBlob testBlob1 = new TestBlob(header, values1);
        row.setBlob1(testBlob1);

        final Header blob = new Header(BLOB_PARAM_TABLE_BLOB_COUNT);
        final byte[] values2 = new byte[BLOB_PARAM_TABLE_BLOB_COUNT];
        for (int i = 0; i < values2.length; ++i)
            values2[i] = (byte)(id + 1);
        final TestBlob testBlob2 = new TestBlob(blob, values2);
        row.setBlob2(testBlob2);

        final byte[] values3 = new byte[BLOB_PARAM_TABLE_HEADER_COUNT];
        for (int i = 0; i < values3.length; ++i)
            values3[i] = (byte)(id + 2);
        final TestBlob testBlob3 = new TestBlob(header, values3);
        row.setBlob3(testBlob3);

        return row;
    }

    private static void checkBlobParamTableRows(List<BlobParamTableRow> rows1, List<BlobParamTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkBlobParamTableRow(rows1.get(i), rows2.get(i));
    }

    private static void checkBlobParamTableRow(BlobParamTableRow row1, BlobParamTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getName(), row2.getName());
        assertEquals(row1.getBlob1(), row2.getBlob1());
        assertEquals(row1.getBlob2(), row2.getBlob2());
        assertEquals(row1.getBlob3(), row2.getBlob3());

        // check reused explicit header parameter
        assertEquals(row2.getBlob1().getBlob(), row2.getBlob3().getBlob());
    }

    private static final int NUM_BLOB_PARAM_TABLE_ROWS = 5;
    private static final int BLOB_PARAM_TABLE_HEADER_COUNT = 10;
    private static final int BLOB_PARAM_TABLE_BLOB_COUNT = 11;
    private static final String FILE_NAME = "explicit_blob_param_test.sqlite";

    private final File file = new File(FILE_NAME);
    private ExplicitParametersDb database = null;
}

File: test/language/explicit_parameters/cpp98/ExplicitBlobParamTest.cpp
#include <cstdio>
#include <vector>
#include <string>

#include "gtest/gtest.h"

#include "zserio/StringConvertUtil.h"

#include "explicit_parameters/ExplicitParametersDb.h"

namespace explicit_parameters
{
namespace explicit_blob_param
{

class ExplicitBlobParamTest : public ::testing::Test
{
public:
    ExplicitBlobParamTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new ExplicitParametersDb(DB_FILE_NAME);
        m_database->createSchema();
        m_header.setCount(BLOB_PARAM_TABLE_HEADER_COUNT);
        m_blob.setCount(BLOB_PARAM_TABLE_BLOB_COUNT);
    }

    ~ExplicitBlobParamTest()
    {
        delete m_database;
    }

protected:
    void fillBlobParamTableRow(BlobParamTableRow& row, uint32_t id, const std::string& name)
    {
        row.setId(id);
        row.setName(name);

        TestBlob testBlob1;
        testBlob1.initialize(m_header);
        zserio::UInt8Array& values1 = testBlob1.getValues();
        for (uint32_t i = 0; i < BLOB_PARAM_TABLE_HEADER_COUNT; ++i)
            values1.push_back(static_cast<uint8_t>(id));
        row.setBlob1(testBlob1);

        TestBlob testBlob2;
        testBlob2.initialize(m_blob);
        zserio::UInt8Array& values2 = testBlob2.getValues();
        for (uint32_t i = 0; i < BLOB_PARAM_TABLE_BLOB_COUNT; ++i)
            values2.push_back(static_cast<uint8_t>(id + 1));
        row.setBlob2(testBlob2);

        TestBlob testBlob3;
        testBlob3.initialize(m_header);
        zserio::UInt8Array& values3 = testBlob3.getValues();
        for (uint32_t i = 0; i < BLOB_PARAM_TABLE_HEADER_COUNT; ++i)
            values3.push_back(static_cast<uint8_t>(id + 2));
        row.setBlob3(testBlob3);
    }

    void fillBlobParamTableRows(std::vector<BlobParamTableRow>& rows)
    {
        rows.clear();
        rows.resize(NUM_BLOB_PARAM_TABLE_ROWS);
        for (uint32_t id = 0; id < NUM_BLOB_PARAM_TABLE_ROWS; ++id)
        {
            const std::string name = "Name" + zserio::convertToString(id);
            fillBlobParamTableRow(rows[id], id, name);
        }
    }

    static void checkBlobParamTableRow(const BlobParamTableRow& row1, const BlobParamTableRow& row2)
    {
        ASSERT_EQ(row1.getId(), row2.getId());
        ASSERT_EQ(row1.getName(), row2.getName());
        ASSERT_EQ(row1.getBlob1(), row2.getBlob1());
        ASSERT_EQ(row1.getBlob2(), row2.getBlob2());
        ASSERT_EQ(row1.getBlob3(), row2.getBlob3());

        // check reused explicit header parameter
        ASSERT_EQ(row2.getBlob1().getBlob(), row2.getBlob3().getBlob());
        // check that even address of the reused explicit header parameter is the same!
        ASSERT_EQ(&row2.getBlob1().getBlob(), &row2.getBlob3().getBlob());
    }

    static void checkBlobParamTableRows(const std::vector<BlobParamTableRow>& rows1,
            const std::vector<BlobParamTableRow>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
            checkBlobParamTableRow(rows1[i], rows2[i]);
    }

    class BlobParamTableParameterProvider : public BlobParamTable::IParameterProvider
    {
    public:
        BlobParamTableParameterProvider()
        {
            m_header.setCount(BLOB_PARAM_TABLE_HEADER_COUNT);
            m_blob.setCount(BLOB_PARAM_TABLE_BLOB_COUNT);
        }

        virtual Header& getHeader(sqlite3_stmt&)
        {
            return m_header;
        }

        virtual Header& getBlob(sqlite3_stmt&)
        {
            return m_blob;
        }

    private:
        Header m_header;
        Header m_blob;
    };

    ExplicitParametersDb* m_database;
    Header m_header;
    Header m_blob;

    static const char DB_FILE_NAME[];

    static const uint32_t NUM_BLOB_PARAM_TABLE_ROWS;
    static const uint32_t BLOB_PARAM_TABLE_HEADER_COUNT;
    static const uint32_t BLOB_PARAM_TABLE_BLOB_COUNT;
};

const char ExplicitBlobParamTest::DB_FILE_NAME[] = "explicit_blob_param_test.sqlite";

const uint32_t ExplicitBlobParamTest::NUM_BLOB_PARAM_TABLE_ROWS = 5;
const uint32_t ExplicitBlobParamTest::BLOB_PARAM_TABLE_HEADER_COUNT = 10;
const uint32_t ExplicitBlobParamTest::BLOB_PARAM_TABLE_BLOB_COUNT = 11;

TEST_F(ExplicitBlobParamTest, readWithoutCondition)
{
    BlobParamTable& blobParamTable = m_database->getBlobParamTable();

    std::vector<BlobParamTableRow> writtenRows;
    fillBlobParamTableRows(writtenRows);
    blobParamTable.write(writtenRows);

    BlobParamTableParameterProvider parameterProvider;
    std::vector<BlobParamTableRow> readRows;
    blobParamTable.read(parameterProvider, readRows);
    checkBlobParamTableRows(writtenRows, readRows);
}

TEST_F(ExplicitBlobParamTest, readWithCondition)
{
    BlobParamTable& blobParamTable = m_database->getBlobParamTable();

    std::vector<BlobParamTableRow> writtenRows;
    fillBlobParamTableRows(writtenRows);
    blobParamTable.write(writtenRows);

    BlobParamTableParameterProvider parameterProvider;
    const std::string condition = "name='Name1'";
    std::vector<BlobParamTableRow> readRows;
    blobParamTable.read(parameterProvider, condition, readRows);
    ASSERT_EQ(1, readRows.size());

    const size_t expectedRowNum = 1;
    checkBlobParamTableRow(writtenRows[expectedRowNum], readRows[0]);
}

TEST_F(ExplicitBlobParamTest, update)
{
    BlobParamTable& blobParamTable = m_database->getBlobParamTable();

    std::vector<BlobParamTableRow> writtenRows;
    fillBlobParamTableRows(writtenRows);
    blobParamTable.write(writtenRows);

    const uint64_t updateRowId = 3;
    BlobParamTableRow updateRow;
    fillBlobParamTableRow(updateRow, updateRowId, "UpdatedName");
    const std::string updateCondition = "id=" + zserio::convertToString(updateRowId);
    blobParamTable.update(updateRow, updateCondition);

    BlobParamTableParameterProvider parameterProvider;
    std::vector<BlobParamTableRow> readRows;
    blobParamTable.read(parameterProvider, updateCondition, readRows);
    ASSERT_EQ(1, readRows.size());

    checkBlobParamTableRow(updateRow, readRows[0]);
}

} // namespace explicit_blob_param
} // namespace explicit_parameters
